### DISCUSSION TITLE
can we check translation key validity (lint or compile)?

### DISCUSSION THREAD

**Discussion by cakasuma** [fnando/i18n-js#651] [Q&A] [answered]
There are many encounters of missing translation key in our team due to either typo or deprecated translation key, and somehow it passed our initial pre-push lint mechanism which then shows undesirable content to the user.

The way we use translation is accessing them by string
```js
i18n.t("some.missing.key")
```

is there a way we could somehow checks if the key put in the translate string is valid?

**Comment by fnando** [comment, marked as answer]
@cakasuma which version of the JavaScript package are you using?

**Reply by cakasuma** [reply]
currently on version 3.1.0

**Reply by fnando** [reply]
You can use something like this in development:

```js
I18n.missingTranslation = function (scope) { 
  throw new Error(`The "${scope}" translation scope is not set`);
};
```

**Reply by fnando** [reply]
btw, the upcoming v4 has `i18n.missingBehavior = "error";`, which would do this for you automatically.

**Reply by desmond990806** [reply]
Hey @fnando, is there any way to "lint" the codebase instead of throwing error when missing translation is called?

**Reply by fnando** [reply]
@desmond990806 we'd need to parse `.[jt]sx?` files to extract static keys (dynamic keys will always be impossible to infer). Given that all tooling for this project is in Ruby, I think this linter should probably be an external tool that uses Node.js and parses files using TypeScript's AST api. Can be done (I have a few scripts parsing files from a project I work on to lint several things like image paths, unused packages, that kinda thing), but I don't know if I want to build it myself.

In case you want to explore it, [this is a nice intro about the subject](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API).

**Reply by fnando** [reply]
@desmond990806 I ended up adding `i18n lint:scripts`, which does what I mentioned above. I'm going to create a beta release soon, so keep an eye on it.

See #677 for details

**Reply by desmond990806** [reply]
@fnando Awesome! Thank you for your work. I'll try on your branch and let you know if I encounter any issues 🎉

**Reply by desmond990806** [reply]
just realized that I couldn't install the branch but will keep an eye on the release, appreciate your work 😄

**Reply by fnando** [reply]
@desmond990806 just uploaded a gem package to https://github.com/fnando/i18n-js/pull/677.

Download the file, unzip it, then install it with `gem install i18n-js-4.1.0.alpha1.gem`. The command you're looking for is `i18n lint:scripts`.

**Reply by desmond990806** [reply]
Hey @fnando , thank you! I will try on it 🙇‍♂️

**Reply by cakasuma** [reply]
hi @fnando can this be incorporated to this package https://github.com/fnando/i18n ?, so we can run this on either yarn/npm 🙏
